### PR TITLE
MH-13534 Add Hourly Data Resolution For Statistics

### PR DESCRIPTION
**Pull request by krinnewitz**
This PR consists of two parts:

The first part allows for different sources for different resolutions with influx. This changes the handling of statistics configurations such that for each provider type, there can be a different configuration format. This makes the mechanism more flexible and reflects the actual variety of options that might come with different sources. This way, we also get rid of the "source" property of the configuration which used to be a compisite string. This didn't make too much sense since we were using JSON already. The new configuration format allows to specify different sources for one influx provider, so that one can use specific measuemrents for certain data resolutions. In practice you might want to use shorter retentions for finer-grained measurements. This change enables you to do so.

This makes it possible to implement the second part: Hourly data resolution for statistics. This adds a new data resolution "hourly" to support finer-grained charts.

Also, I have included two sample configs for an easy review. I will remove these configs before this PR gets merged.

This work is sponsored by SWITCH.

### CODE REVIEW

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java [hunk 1]
```diff
@@ -67,6 +68,12 @@ private StatisticsUtil() {
           result.add(currentWeekStart); // No min, max here, because influx uses previous monday as bucket label
           localStart = localStart.plusDays(7).plusDays(daysOff).withHour(0).withMinute(0).withSecond(0).withNano(0);
           break;
+        case HOURLY:
+          final Instant currentHourStart = ZonedDateTime.of(localStart.withMinute(0).withSecond(0).withNano(0), zoneId)
+              .toInstant();
+          result.add(Ordering.natural().min(to, Ordering.natural().max(from, currentHourStart)));
+          localStart = localStart.plusHours(1).withMinute(0).withSecond(0).withNano(0);
+          break;
```
Comment (KatrinIhler) [inline modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java:76]: Do we need the min here? I assume that neither from nor currentHourStart can ever be > to, but I might be wrong.

File: modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java [hunk 2]
```diff
@@ -49,6 +49,31 @@ private static Instant randomInstant() {
...
+      assertEquals(
+          LocalDateTime.ofInstant(
+              to,
+              ZoneOffset.UTC)
+              .withMinute(0)
+              .withSecond(0)
+              .withNano(0)
+              .toInstant(ZoneOffset.UTC),
+          buckets.get(buckets.size() - 1));
+    }
```
Comment (KatrinIhler) [inline modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java:92]: This test confuses me.
1. Why do we test with random instants instead of a set of predefined test cases?
2. Why do we only check the last bucket?

File: docs/guides/admin/docs/configuration/admin-ui/statistics.md [hunk 3]
```diff
@@ -94,13 +90,18 @@ Here is an example json configuration for a provider which generates charts for
...
+    "aggregation": "SUM",
+    "aggregationVariable": "value",
+    "resourceIdName": "episodeId",
+    "resolutions": [
+      "DAILY",
+      "WEEKLY",
+      "MONTHLY",
+      "YEARLY"
+    ]
+  }],
```
Comment (KatrinIhler) [inline docs/guides/admin/docs/configuration/admin-ui/statistics.md:104]: Looks much better now! :)

**Review by KatrinIhler — COMMENTED** [review]
My thoughts after looking the code over:

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java (hunk 1 above)
Comment (krinnewitz) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java:76]: Here we set `currentHourStart` to  an Instant with minute `0`. The idea of the min max is to not run out of the bounds defined by `from` and `to` in case `from` and/or `to` have minute != 0. E.g. in case of from=...T08:30:00Z we don't want to get the 08:00 Instant. However, this is more or less theoretical since the frontend should only provide full-hour Instants. (I am not sure what happens in the greatly feared timezone "America/St_Johns", though...)

File: modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java (hunk 2 above)
Comment (krinnewitz) [reply modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java:92]: We also check the first bucket, but not the ones in between :) Using the random Instants was easier than sitting down an thinking about all possible edge cases. 

So the answer to both of your questions is: laziness (or "prioritization" in other words :)). I will spend some extra time to improve the tests.

Comment (krinnewitz) [reply modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java:92]: I have extended the tests a bit.

Comment (KatrinIhler) [reply modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java:92]: Alright, now I'm wondering why we need the randomised tests at all if we now have a nice set of predefined instants to test. I'm not too fond of random test cases because this could lead to tests failing only sometimes, which is terrible for debugging...

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java (hunk 1 above)
Comment (KatrinIhler) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java:76]: My thought process is something like this: We make sure that localStart is before localEnd in the loop, and currentHourStart can only ever become less by setting minutes, seconds etc to 0, not more, right? So while currentHourStart could become less than from, I can't imagine a case where currentHourStart could ever be bigger than to. Or am I missing something?

File: modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java (hunk 2 above)
Comment (krinnewitz) [reply modules/statistics-service-api/src/test/java/org/opencastproject/statistics/api/StatisticsUtilTest.java:92]: Well, I wouldn't consider 3 date pairs a "nice set of predefined instants to test". The idea with the random tests is to log the input so you can debug it very easily. However, I just noticed that Opencast has set the log level to "WARN" for all tests running with maven so my logs don't get printed. Therefore, I added the from and to dates to the output which is printed in case a test fails.

I actually like randomized testdata since most of the time, people just test the cases they can imagine (which are mostly identical with the ones they have in mind when developing).

Though it still makes sense to have hand-picked test data as well to make sure that some important cases are always tested.

File: modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java (hunk 1 above)
Comment (krinnewitz) [reply modules/statistics-service-api/src/main/java/org/opencastproject/statistics/api/StatisticsUtil.java:76]: Ah now I see it. You are right. `min` is not required, but `max` is. I removed the `min`.

**Review by KatrinIhler — APPROVED** [review]

**Review by wsmirnow — APPROVED** [review]
I'm fine with this PR. Thank you.
